### PR TITLE
FTP and alarms

### PR DESCRIPTION
**Pull request by regiskuckaertz**
I've made a few structural changes that will hopefully address the file issues NLA have experienced once and for all, or in any case help us diagnose the problem. All the changes are in the lambda sending the file via FTP. That lambda is triggered by an S3 event, whenever the another lambda kicks off the Athena request and the latter outputs the result to the bucket.

Summary of changes:
- on start, sends the size of the CSV over to CloudWatch
- adds an alarm that triggers whenever that size goes below 100MB
- updates packages to their latest version
- changes the FTP package to one that has been updated more recently (the one we were using dates back 4 years)
- uploads the generated ZIP to an S3 bucket: this will be useful to check if the file we generated is corrupt or not, if that ever happens.
- changes the way handling the end of the FTP session is done. Previously the lambda was fulfilled as soon as we called `ftpClient.end()`, but it turns out that API call will not close the connection: it merely tells the clients to close the connection _once all the remaining operations have completed_. To that, I've updated the logic to listen the `end` and `close` events, the latter of which becomes the signal the lambda is fulfilled.

### CODE REVIEW

File: ftp-upload/src/lambda.ts [hunk 1]
```diff
@@ -1,13 +1,14 @@
 import {Config} from './config';
-import {Readable} from "stream";
 
 const AWS = require('aws-sdk');
-const ftp = require('ftp');
+const FtpClient = require('ftp');
```
Comment (JustinPinner) [inline ftp-upload/src/lambda.ts]: The 'ftp' required is exposed by `@icetee/ftp` dependency I presume? By this I mean that the `ftp` reference is still valid to `require` as-is?

**Review by JustinPinner — APPROVED** [review]
It'll be good to see what we catch in the S3 bucket! I had one noob question, but otherwise 👍

File: ftp-upload/src/lambda.ts (hunk 1 above)
Comment (regiskuckaertz) [reply ftp-upload/src/lambda.ts]: Good point!

Comment (regiskuckaertz) [reply ftp-upload/src/lambda.ts]: The build didn't complain but I prefer to avoid any ambiguity.

**Comment by regiskuckaertz** [comment]
Alright. I've made several changes in the last few commits that fall in two categories:
- misconfigured API calls: I had made a few mistakes in the JS apis for CloudWatch and S3
- allow local testing: assuming a role doesn't work with dev credentials, which is why I was unable to run my lambda locally 😭 so I removed that, now it only runs in the Lambda itself.

File: ftp-upload/cloudformation.yaml [hunk 2]
```diff
@@ -88,4 +95,25 @@ Resources:
...
+      Namespace: AWS/Lambda
+      MetricName: SizeOfCSV
+      Dimensions:
+        - Name: FunctionName
+          Value:
+            !Ref Lambda
+      Statistic: Sum
+      ComparisonOperator: LessThanThreshold
+      Threshold: 100
+      Unit: Megabytes
```
Comment (JustinPinner) [inline ftp-upload/cloudformation.yaml:114]: Blimey, these must be pretty big files usually then?

Comment (regiskuckaertz) [reply ftp-upload/cloudformation.yaml:114]: yeah around 170–180, here I just wanted to be conservative

File: ftp-upload/src/lambda.ts [hunk 3]
```diff
@@ -53,104 +52,164 @@ async function run(event) {
...
+        output.on('close', () => {
+            const fileSizeInMB = archive.pointer()/1024/1024;
+            console.log(`Finished zipping CSV file to ${outputFile} (${fileSizeInMB.toFixed(2)}MB)`);
+            resolve(outputFile);
+        });
+
+        // good practice to catch warnings (ie stat failures and other non-blocking errors)
+        archive.on('warning', (err) => {
+            if (err.code === 'ENOENT') {
+                console.warn("Woopsie, something weird happened", err);
```
Comment (JustinPinner) [inline ftp-upload/src/lambda.ts]: should we `reject(err)` here, too?

**Review by JustinPinner — APPROVED** [review]
This looks good as far as I can see. It's hard to do a side-by-side comparison but the new code looks sensible 👍

**Comment by regiskuckaertz** [comment]
Yeah, sorry about that. Next time in Scala ⚡️

**Comment by regiskuckaertz** [comment]
I'll merge on Monday.

**Pull request merged**